exporter/containerimage/writer.go

```diff
@@ -333,6 +385,132 @@ func (ic *ImageWriter) commitDistributionManifest(ctx context.Context, ref cache
 	}, &configDesc, nil
 }
 
+func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, p exptypes.Platform, target string, oci bool, attestations []exporter.Attestation, inTotos []InTotoAttestation) (*ocispecs.Descriptor, *ocispecs.Descriptor, error) {
+	var (
+		attestationType                = exptypes.MediaTypeAttestationType
+		manifestType                   = ocispecs.MediaTypeImageManifest
+		configType                     = ocispecs.MediaTypeImageConfig
+		annotationReferenceType        = exptypes.OCIAnnotationReferenceType
+		annotationReferenceDigest      = exptypes.OCIAnnotationReferenceDigest
+		annotationReferenceDescription = exptypes.OCIAnnotationReferenceDescription
+	)
+	if !oci {
+		attestationType = exptypes.MediaTypeDockerSchema2AttestationType
+		manifestType = images.MediaTypeDockerSchema2Manifest
+		configType = images.MediaTypeDockerSchema2Config
+		annotationReferenceType = exptypes.DockerAnnotationReferenceType
+		annotationReferenceDigest = exptypes.DockerAnnotationReferenceDigest
+		annotationReferenceDescription = exptypes.DockerAnnotationReferenceDescription
+	}
+
+	layers := []ocispecs.Descriptor{}
+	for _, inToto := range inTotos {
+		data, err := json.Marshal(inToto)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		digest := digest.FromBytes(data)
+		desc := ocispecs.Descriptor{
+			MediaType: attestationType,
+			Digest:    digest,
+			Size:      int64(len(data)),
+			Annotations: map[string]string{
+				"containerd.io/uncompressed": digest.String(),
+				"in-toto.io/predicate-type":  inToto.PredicateType,
```

Comment:
So this implies poinging at individual attestations, is that somehow essential in the context of propsal F?

From package-checker perspective, there are many attestions of 5 different predicate types, it seems natural to just store them in one tarball and referene as a bundle, possibly together with some other attestation. More broadly, I'm thinking that if all of attestation related to a DOI could be bundled together, that woild make it easier to retrieve.

In a way, I guess my point is really weather predicate type should be even defined at this level? Perhaps it'd be better to avoid any assumption about the format at this level?
I guess I'm not even sure if buildkit needs to be aware of intoto at all and do any serialisation etc, it could just handle opaque documents or document bundles instead...

Reply:
The predicate type being defined here was something that came up in discussion with @tonistiigi - we need a way for tooling to determine what an attestation is, preferably before pulling it. For example, a tool might want to check signatures, but not pull SBOMs, or vice-versa. The more homogeneous and standardized the format, the easier it will be to work with and program with.

I think there needs to be *some* level of understanding of the attestation format, at least as far as the separation of "what type the attestation is" (specified in the api), "what content the attestation has" (specified as content at a ref) and "what the attestation is about" (which may be manually specified or automatically determined as the final image result).

I think Proposal F is an implementation detail here - it's not fixed in stone, and future exporters might want to export differently. Having the data split out transparently instead of in an opaque bundle gives us some more flexibility here in the future, if we decide to change the model.

Reply:
Yeah I tend to think that you may well want to bundle a set of predicates together. I do agree that knowing what they are up front is useful but maybe we could have a collection type ie "predicates about SBOMs" rather than being specific?

Reply:
This manifest is already a bundle. If the format needs extra groups they can just have their own manifest.

If we also want optional tarball mediatype in here then that isn't that hard as well. Basically, just change the refpath to array. I'm not sure what the semantics for the filenames inside that tarball should be then though as afaik there is no tooling for such formats atm.

Reply:
> To begin with, we will need to have payload type 

That's the mediatype field

> I think names could really be anything, doesn't matter all that much. The existence of tooling is irrelevant, this is not a particularly mature teritory, there is plenty of gaps.

We still need something that the potential tooling could use a specification for the tooling for how to do it. All changes in this level are breaking for any tool using it.